File: Java-for-Collage/sjf.java
                                    /*Implementation of SJF by Ayush Pal */  /* */
/*Objective:-Implementation of Shortest job first Cpu scheduling algorithum */
/*Description:-In shortest job first the process which has shortest job will get cpu first.The program consist of two things to input given by the use that is arrival time and burst 
 * time of process.The ouput of program displays a table that consist of coloumn of process number, arrival time
 * burst time, completion time, turnaround time, waiting time. The program also displays average waiting time and 
 * average turnaround time of processes */

import java.util.Scanner;

public class SJF {
    public static void main(String args[])
    {
        Scanner sc = new Scanner(System.in);
        /*use of scanner class to enter data */
        System.out.println ("Please enter no of process:");
        //tell number of process
        int n = sc.nextInt();
        //process id array
        int pid[] = new int[n];
        //process burst time array
        int at[] = new int[n];
        //process arrival time array
        int bt[] = new int[n];
        //process completion time
        int ct[] = new int[n];
        //process turnaround time array
        int ta[] = new int[n];
        //process waiting time array
        int wt[] = new int[n];
        /*this array acts a flag that consist of 0,1 in which 0 is for incomplete and 1 for complete  */
        int f[] = new int[n];
        /*st stands for start time if the process starts from one the it will be 1 or same as arrival time but intilization
         * is 0
         * tot variable to use check if we have finished all the process 
        */
        int st=0, tot=0;
        /*intilize of average waiting and turnaround time to 0 */
        float avgwt=0, avgta=0;
        //taking input from the user
        for(int i=0;i<n;i++)
        {
            System.out.println ("Please enter process " + (i+1) + " arrival time:");
            at[i] = sc.nextInt();
            System.out.println ("Please enter process " + (i+1) + " brust time:");
            bt[i] = sc.nextInt();
            pid[i] = i+1;
            f[i] = 0;
        }


        while(true)
        {
            int c=n, min = 999;

            /*if the tot is equal to n that is number of process then we have done our job and it is finshed so we break out of while
             * loop
             */
            if (tot == n){
                break;}
            /*If not equal to number of process */    

            for (int i=0; i<n; i++)
            {
              /*We will check which process has least burst time and has arrived first */
                if ((at[i] <= st) && (f[i] == 0) && (bt[i]<min))
                {
                /*if statement will execute only if arrival time is less than starting time 
                 * and flag is 0 that is it is not complete and burst time must be min we can take any number in 
                 * place of min as first process must be smallest
                 */
                    min=bt[i];
                    c=i;
                }
            }
            /*what if we dont have any process arrived at time 0 we should increase the start time */
            if (c==n){
                st++;}
            else
            {
                //calculation of completion time
                ct[c]=st+bt[c];
                //we will increase the stop time by burst time
                st+=bt[c];
                //calculation of turnaround time
                ta[c]=ct[c]-at[c];
                //calculation of waiting time
                wt[c]=ta[c]-bt[c];
                //raising the flag to one
                f[c]=1;
                //getting the new process id
                pid[tot] = c + 1;
                //increase the tot value to 1 and it will work until tot=number of process
                tot++;
            }
        }

        System.out.println("Process\t\tAT\t\tBT\t\tCT\t\tTAT\t\tWT");
        /* Calculation of average waiting time and turnaround time*/
        for(int i=0;i<n;i++)
        {
            avgwt+= wt[i];
            avgta+= ta[i];
            System.out.println(pid[i]+"\t\t"+at[i]+"\t\t"+bt[i]+"\t\t"+ct[i]+"\t\t"+ta[i]+"\t\t"+wt[i]);
        }
        System.out.println ("\naverage tat is "+ (float)(avgta/n)+"unit time");
        System.out.println ("average wt is "+ (float)(avgwt/n)+"unit time");
        System.out.println("Where AT is arrival time,BT is burst time, CT is compltion time, TAT is turnaround time, WT is waiting time in above table");
        sc.close();
    }
}

File: Operating System/sjf.c
/*
Objective - Implementation of Shortest job first Cpu scheduling algorithum
Shoetest Job first is a cpu scheduling algorithm which has least waiting time 
Arguments:
    1-Arrival Time
    2-Burst Time
Returns:
    1:Average Waiting Time
    2:Average Turnaround TIme
*/
#include <stdio.h>

// Function to sort processes by burst time (for SJF)
void sortProcessesByBurstTime(int n, int processes[], int bt[], int at[]) {
    for (int i = 0; i < n - 1; i++) {
        for (int j = i + 1; j < n; j++) {
            if (bt[i] > bt[j]) {
                // Swap burst times
                int temp = bt[i];
                bt[i] = bt[j];
                bt[j] = temp;

                // Swap process IDs
                temp = processes[i];
                processes[i] = processes[j];
                processes[j] = temp;

                // Swap arrival times
                temp = at[i];
                at[i] = at[j];
                at[j] = temp;
            }
        }
    }
}

// Function to calculate waiting time for each process
void waitingTime(int processes[], int n, int bt[], int at[], int wt[]) {
    int service_time[n]; // To store the cumulative service time
    service_time[0] = at[0];
    wt[0] = 0;  // Waiting time for first process is 0

    for (int i = 1; i < n; i++) {
        // Add burst time of the previous process to its service time
        service_time[i] = service_time[i - 1] + bt[i - 1];

        // Calculate waiting time as the difference between service time and arrival time
        wt[i] = service_time[i] - at[i];

        // If waiting time is negative, set it to 0
        if (wt[i] < 0) {
            wt[i] = 0;
        }
    }
}

// Function to calculate turnaround time for each process
void TurnaroundTime(int processes[], int n, int bt[], int wt[], int tat[]) {
    for (int i = 0; i < n; i++) {
        tat[i] = bt[i] + wt[i];  // Turnaround time = burst time + waiting time
    }
}

// Function to calculate average waiting time
void AvgwaitingTime(int processes[], int n, int wt[]) {
    int total_wt = 0;
    for (int i = 0; i < n; i++) {
        total_wt += wt[i];
    }
    printf("Average Waiting Time = %.2f\n", (float)total_wt / n);
}

// Function to calculate average turnaround time
void AverageTAT(int processes[], int n, int tat[]) {
    int total_tat = 0;
    for (int i = 0; i < n; i++) {
        total_tat += tat[i];
    }
    printf("Average Turnaround Time = %.2f\n", (float)total_tat / n);
}

int main() {
    int n;

    printf("Enter the number of processes: ");
    scanf("%d", &n);

    int processes[n], bt[n], at[n], wt[n], tat[n];

    // Input burst times and arrival times
    printf("Enter burst time and arrival time for each process:\n");
    for (int i = 0; i < n; i++) {
        printf("Process %d:Arrival Time | Burst Time: ", i + 1);
        scanf("%d %d", &at[i], &bt[i]);
        processes[i] = i + 1;
    }

    // Sort processes by burst time (Shortest Job First)
    sortProcessesByBurstTime(n, processes, bt, at);

    // Calculate waiting time and turnaround time
    waitingTime(processes, n, bt, at, wt);
    TurnaroundTime(processes, n, bt, wt, tat);

    // Display process information
    printf("\nProcess\tArrival Time\tBurst Time\tWaiting Time\tTurnaround Time\n");
    for (int i = 0; i < n; i++) {
        printf("%d\t%d\t\t%d\t\t%d\t\t%d\n", processes[i], at[i], bt[i], wt[i], tat[i]);
    }

    // Calculate and display average waiting time and turnaround time
    AvgwaitingTime(processes, n, wt);
    AverageTAT(processes, n, tat);

    return 0;
}